### ISSUE TITLE
cannot compile on ubuntu 14.04 / ruby 2.2.3

### ISSUE THREAD

**Issue by masterkain** [copiousfreetime/hitimes#48]
on my osx local machine everything's working fine, I recently upgraded to ruby 2.2.3 and reinstalled the gem with no trouble.

however on my production server this gem cannot be installed, builds on 2.2.1 but it does not build on 2.2.3:

```
/home/ubuntu/.rvm/rubies/ruby-2.2.3/bin/ruby -r ./siteconf20150917-1776-wbf434.rb extconf.rb
checking for clock_gettime() in -lrt... *** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
	--with-opt-dir
	--without-opt-dir
	--with-opt-include
	--without-opt-include=${opt-dir}/include
	--with-opt-lib
	--without-opt-lib=${opt-dir}/lib
	--with-make-prog
	--without-make-prog
	--srcdir=.
	--curdir
	--ruby=/home/ubuntu/.rvm/rubies/ruby-2.2.3/bin/$(RUBY_BASE_NAME)
	--with-rtlib
	--without-rtlib
/home/ubuntu/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/mkmf.rb:456:in `try_do': The compiler failed to generate an executable file. (RuntimeError)
You have to install development tools first.
	from /home/ubuntu/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/mkmf.rb:541:in `try_link0'
	from /home/ubuntu/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/mkmf.rb:556:in `try_link'
	from /home/ubuntu/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/mkmf.rb:735:in `try_func'
	from /home/ubuntu/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/mkmf.rb:966:in `block in have_library'
	from /home/ubuntu/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/mkmf.rb:911:in `block in checking_for'
	from /home/ubuntu/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/mkmf.rb:351:in `block (2 levels) in postpone'
	from /home/ubuntu/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/mkmf.rb:321:in `open'
	from /home/ubuntu/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/mkmf.rb:351:in `block in postpone'
	from /home/ubuntu/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/mkmf.rb:321:in `open'
	from /home/ubuntu/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/mkmf.rb:347:in `postpone'
	from /home/ubuntu/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/mkmf.rb:910:in `checking_for'
	from /home/ubuntu/.rvm/rubies/ruby-2.2.3/lib/ruby/2.2.0/mkmf.rb:961:in `have_library'
	from extconf.rb:10:in `<main>'

extconf failed, exit code 1
```

**Comment by copiousfreetime** [comment]
That is pretty odd, can you double check a couple of things first? The error message `You have to install development tools first` may indicate that not all the build chain is setup. Easiest way to get that is with `apt-get install build-essential`.  Since you say it installs fine on ruby 2.2.1 I'm assuming that you already have build-essential installed. Just covering all the bases.

If you already have that installed then can you upload the contents of the `mkmf.log` file? If hitimes 1.2.3  builds and installs just fine on ruby 2.2.1 on the same server that you are upgrading to ruby 2.2.3 then my only thoughts on why it wouldn't build for ruby 2.2.3 is that ruby 2.2.3 install is not quite right. The info in the mkmf.log file should help point us in the right direction.

**Comment by oree** [comment]
Same error on Ubuntu 14.04 / Ruby 2.1.7!

Then again, works on Ubuntu 12.04 / Ruby 2.1.7...

**Comment by oree** [comment]
The error in mkmf.log:

<pre>
"gcc -o conftest -I/home/deploy/.rvm/rubies/ruby-2.1.7/include/ruby-2.1.0/x86_64-linux -I/home/deploy/.rvm/rubies/ruby-2.1.7/include/ruby-2.1.0/ruby/backward -I/home/deploy/.rvm/rubies/ruby-2.1.7
/include/ruby-2.1.0 -I.     -O3 -fno-fast-math -ggdb3 -Wall -Wextra -Wno-unused-parameter -Wno-parentheses -Wno-long-long -Wno-missing-field-initializers -Wunused-variable -Wpointer-arith -Wwrite
-strings -Wdeclaration-after-statement -Wimplicit-function-declaration  -fPIC conftest.c  -L. -L/home/deploy/.rvm/rubies/ruby-2.1.7/lib -Wl,-R/home/deploy/.rvm/rubies/ruby-2.1.7/lib -L. -fstack-p
rotector -rdynamic -Wl,-export-dynamic     -Wl,-rpath,'/../lib' -Wl,-R -Wl,'/../lib' -L'/../lib' -lruby  -lpthread -lgmp -ldl -lcrypt -lm   -lc"
/usr/bin/ld: cannot find -lgmp
collect2: error: ld returned 1 exit status
checked program was:
/* begin */
1: #include "ruby.h"
2: 
3: int main(int argc, char **argv)
4: {
5:   return 0;
6: }
/* end */
</pre>

**Comment by oree** [comment]
apt-get install libgmp-dev

fixed the problem

**Comment by dnesteryuk** [comment]
@oree thank you for sharing the solution!!!

**Comment by copiousfreetime** [comment]
@oree thanks, good to know, I'll see if I can add some additional output during build to provide a more useful output.

**Comment by IvanTashkinov** [comment]
@oree, :+1:

**Comment by dcorlieu** [comment]
Thank you, oree! Been searching for hours!

**Comment by copiousfreetime** [comment]
The `-lgmp` requirement is not part of hitimes itself, it is part of ruby. You can see this by executing `ruby -rrbconfig -e 'puts RbConfig::CONFIG.find_all { |k,v| puts "#{k} : #{v}" if v =~ /gmp/ }'`.

What I'm guessing is that the build of ruby that folks are using is probably a binary install that was built against libgmp.

When a ruby extension is installed the compilation flags used on the extension are the same that are used with the original build of ruby. In this case, the original build of ruby was built against libgmp and then either `libgmp-dev` was removed, or `libgmp-dev` was never installed to begin with.

Unfortunately since this appears to be a ruby installation issue, I don't think I can do anything to help it. I would almost expect this same error to appear on any ruby gem that has an extension.

**Comment by sinhtruong** [comment]
@oree thanks very much (y)

**Comment by dhanabalt** [comment]
@oree  Thanks... saved my time

**Comment by nrowegt** [comment]
@oree thanks man, that did the trick.

**Comment by vans88** [comment]
@masterkain  Many thanks for your comments, I fixed my error use the same method.

**Comment by mkarlovich** [comment]
@oree that also fixed issue running on Windows 10 Ubuntu 16.04 / Ruby 2.3.3.  Thanks!
